### ISSUE TITLE
Strange behavioral from 'std::env' when calling Rust program from C++ Program

### ISSUE THREAD

**Issue by PooyaEimandar** [rust-lang/rust#88108]
I port an application from C++ to RUST which is responsible to parse the input arguments and then do some actions. 
I tried to call this RUST application from another C++ application via [CreateProcess ](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa) function. 
But something strange happened, I think the lifetime of passing parameters in C++ affects on Rust program.
My rust program receives the arguments in C++ Debug mode successfully but it could not received same arguments in Release mode of C++.
I think std::env of Rust should work isolated from parent process, am I right?

Related issue [#44650](https://github.com/rust-lang/rust/issues/44650) 

`rustc --version --verbose`:
```
rustc 1.54.0 (a178d0322 2021-07-26)
binary: rustc
commit-hash: a178d0322ce20e33eac124758e837cbd80a6f633
commit-date: 2021-07-26
host: x86_64-pc-windows-msvc
release: 1.54.0
LLVM version: 12.0.1
```

**Comment by ChrisDenton** [comment]
Yes, processes are separate. Would it be possible to provide an example of this bug? I mean a small C++ program and a small Rust program that shows the problem. This would make it easier to investigate the problem.
